### ISSUE TITLE
Random forests work well, but prioritize some strange metrics

### ISSUE THREAD

**Issue by wkumler** [wkumler/MS_metrics#5]
We've transitioned into using random forests after struggling with clustering. Random forests have some advantages, primarily interpretability, invariance with scaling, ability to handle non-linear behavior, and returns the relative importance of each variable. Plus, they also don't require insane amounts of data and we have at most 2,040 features because I refuse to classify any more in this data set.

  - Interpretability: you get clear success/failure metrics based on distinct cuts in the feature space
  - Invariance with scaling: clustering can only really handle things if they're on relatively similar scales because the distance matrix is calculated identically between different metrics. Random forests can cut on any scale which means we sidestep the scaling question.
  - Ability to handle non-linear behavior: I'm worried about metrics like peak width where there should be a sweet spot, with peaks too narrow or too wide will be problematic. Random forests can cut twice on the same metric, once high and once low, and isolate that middle area.
  - Returns the relative importance of the variables! Tells you which metrics are most useful.

If I were to design a random forest kinda manually, I would guess at the following cuts. Good peaks should have:
  - Peak width less than 60 and above 30(?)
  - High med_SNR (high signal-to-noise)
  - High med_cor (good peak shape)
  - Present in lots of files
  - Retention time greater than 1 (avoids weird spiky ones super early) and probably less than 13-14(?) to avoid solvent peak later
  - Maybe even cuts out retention time in the middle around the sulfate peak?
  - Mean_mz less than maybe 300? I don't remember seeing any good peaks above that mass
  - Low rt_sd (consistent retention times)
  - Large mean_area
  - High shape_cor and area cor (has an isotope)
  - Anything with a large t_pval

I'm planning to use the `randomForest` library because that's what most tutorials use and maybe `rpart` for single decision trees.

**Comment by wkumler** [comment]
I'm getting some weird results from the default xcms parameters where `scale` and `f` are weirdly powerful. Here's a quick example of this using all the features:

```r
set.seed(123)
traintestlist <- features_extracted %>%
  filter(feat_class%in%c("Good", "Bad")) %>%
  slice_sample(n = nrow(.)) %>%
  split(rbernoulli(nrow(.), 0.2)) %>%
  setNames(c("train", "test"))
rfmodel <- traintestlist$train %>%
  select(-feat_id, -feat_class) %>%
  data.matrix() %>%
  randomForest(y=factor(traintestlist$train$feat_class))
rfmodel
varImpPlot(rfmodel)

traintestlist$test %>%
  select(-feat_id, -feat_class) %>%
  data.matrix() %>%
  predict(object=rfmodel) %>%
  as.data.frame() %>%
  setNames("pred_class") %>%
  cbind(traintestlist$test) %>%
  with(table(pred_class, feat_class))
```

which performs pretty well on both the training set:
Confusion matrix:
|        | Bad | Good |
| --- | --- | --- |
| Bad | 1214 | 4 |
| Good | 26 | 158 |

and the test set:
|        | Bad | Good |
| --- | --- | --- |
| Bad | 316 | 6 |
| Good | 2 | 48 |

but the feature importance is not at ALL what I expected:

![image](https://user-images.githubusercontent.com/43116923/216736001-47ef0f42-e66c-4b52-899e-cb9aa0d2c08b.png)

**Comment by wkumler** [comment]
I have no idea why these parameters would be so strong. 

As far as I can tell, the `scale` parameter is the wavelet scale at which the wavelet coefficients reached the highest value. This also turns out to be 10 for most of them, strangely. `extendLengthMSW` is enabled so that's not the explanation, but I can't figure out why so many peaks have a "best" scale right close to 10. 

The `f` parameter is even weirder. This is literally just the counter in a `for` loop as the algorithm iterates over various regions of interest:

```
for (f in 1:length(roiList) {
        feat <- roiList[[f]]
        ...
        peaks <- rbind(peaks, c(mzmean, mzrange, 
                                NA, NA, NA, NA, NA, maxint, round((maxint - 
                                  baseline)/sdnoise), NA, NA, NA, NA, 
                                f, dppm, best.scale, td[best.scale.pos], 
                                td[lwpos], td[rwpos], NA, NA))
}
```

and I have no idea what values it's cutting on or how it's using this parameter to decide what peaks are good. Neither of these parameters look especially powerful in separating good from bad:

![image](https://user-images.githubusercontent.com/43116923/217334997-581c2984-5dc4-4702-8706-41faa35b9ac7.png)

![image](https://user-images.githubusercontent.com/43116923/217335080-ad5edf4a-1be0-4c99-96d7-6e82827ef308.png)

especially relative to the really strong features like med_cor:

![image](https://user-images.githubusercontent.com/43116923/217336827-2828b40b-b804-4297-b154-73bb443d73d8.png)

and med_SNR:

![image](https://user-images.githubusercontent.com/43116923/217337013-c0fd8e78-cd71-40ef-abaa-1707a86420e8.png)

I wonder how a manually-created decision tree using the original expectations would perform?
